client/base-rtc.js

```diff
@@ -0,0 +1,335 @@
+    pc.onicecandidate = function (event) {
+      if (event.candidate !== null) {
+        this._send({ 'ice': event.candidate }, remoteUser, 'ice');
+      }
+    }.bind(this);
+
+    // Add our media stream if we have one
+    if (this.localStream) {
+      pc.addStream(this.localStream);
+    }
+
+    // finally, add new peer connection to our cache
+    this.peerConnections[remoteUser] = pc;
+    return pc;
+  };
+
+  // Subclasses must implement -- some will connect to a room, others to a user.
+  // In either case, the room is the room they are connecting to. It takes an 
+  // object of the form: { name: 'living room', presenter: 'fred' }. The user
+  // is the user they are connecting as (not who they are connecting to).
+  baseRTC.connect = function (room, user) {
+    console.log('ERROR: called connect method on base class. Subclasses must implement .connect');
+  };
+
+  baseRTC.disconnect = function (room, user) {
+    console.log('ERROR: called connect method on base class. Subclasses must implement .connect');
+  };
+
+  // Ask all clients in room to send an offer. They all generate offers that
+  // we can then give an answer to. Why do this? Because we don't know who else
+  // is in the room, so we don't know what peer connections to generate so we can 
+  // create offers. So instead, we ask for offers and now we know which peer 
+  // connections to create.
+  // TODO: fix the hacky way we're sending a message to the entire room. It shouldn't go to 'everyone'
+  baseRTC.connectToRoom = function (room) {
+    this._send({ 'requester': this.me }, 'everyone', 'request-for-offer'); 
+  };
+
+  // Create peer connection to a specific user and send them an offer. 
+  baseRTC.connectToUser = function (remoteUser) {
+    var pc = this.createPeerConnection(remoteUser);
+    // createOffer method generates an offer (SDP session description) that will be sent to peer.
+    pc.createOffer(function(description) {
+      // Before the offer is sent to the peer, the local description must be set to match the offer.
+      pc.setLocalDescription(description, function () {
+        // The offer is stringified utilizing the _send helper method below.
+        this._send(description, remoteUser, 'offer'); 
+      }.bind(this));
+    }.bind(this), function (err) { console.log('offer erorr: ', err); });
+  };
+
+  // Close peer connection to a specific user.
+  // This method gets called by an audience member to turn off their mic and stop audio from streaming 
+  baseRTC.disconnectFromUser = function(remoteUser){
+    // Lookup user in peerConnections object;
+    var pc = this.peerConnections[remoteUser];
+    // Utilize built-in RTCPeerConnection close method https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/close
+    pc.removeStream(this.localStream);
+    // pc.close();
+  };
```

Comment:
I wasted a bunch of time trying to get the onremovestream handler working.  If you notice in the presenter controller, i have both an onremovestream and an oniceconnectionstatechange.  I've learned that onremovestream does not get fired until the entire handshake is renegotiated.

http://stackoverflow.com/questions/16478486/webrtc-function-removestream-dont-launch-event-onremovestream-javascript

I couldn't find a graceful way to do this with our current structure (would take a lot of refactoring from what I saw).  The best that I could come up with was to utilize connect again but that fires up a new video immediately after, obviously, which is not what we want by turning the mic off.